### PR TITLE
[Compute] Fix #20174: `az vm create`: Determine plan information when using image alias

### LINKED ISSUE #20174: `az vm create` fails for Flatcar distro alias because plan information is unset
## Describe the bug

**Command Name**
`az vm create --image Flatcar`

**Errors:**
```
{"status":"Failed","error":{"code":"DeploymentFailed","message":"At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/DeployOperations for usage details.","details":[{"code":"Conflict","message":"{\r\n  \"status\": \"Failed\",\r\n  \"error\": {\r\n    \"code\": \"ResourceDeploymentFailure\",\r\n    \"message\": \"The resource operation completed with terminal provisioning state 'Failed'.\",\r\n    \"details\": [\r\n      {\r\n        \"code\": \"VMMarketplaceInvalidInput\",\r\n        \"message\": \"Creating a virtual machine from Marketplace image or a custom image sourced from a Marketplace image requires Plan information in the request. VM: '/subscriptions/47635d02-50bb-4f1f-8b44-e9e9518015e6/resourceGroups/group7088/providers/Microsoft.Compute/virtualMachines/flatcar-k8s-001'.\"\r\n      }\r\n    ]\r\n  }\r\n}"}]}}

```

## To Reproduce:
Steps to reproduce the behavior. Note that argument values have been redacted, as they may contain sensitive information.

- _Put any pre-requisite steps here..._
- `az vm create --image Flatcar --size {} -g {} -n {} --storage-sku {} --public-ip-sku {} --admin-username {}`

## Expected Behavior

Expected success. `az vm create --image kinvolk:flatcar-container-linux-free:stable:latest --size {} -g {} -n {}` works correctly, so this looks to be a client-side issue.

## Environment Summary
```
Linux-5.4.0-1062-azure-x86_64-with-debian-10.2 (Cloud Shell), Common Base Linux Delridge (quinault)
Python 3.6.10
Installer: DEB

azure-cli 2.29.0 *

Extensions:
ai-examples 0.2.5
ssh 0.1.6

```

### PR DESCRIPTION
**Pull request by jepio**
**Description**

Some image aliases require "plan information" when used. The "Flatcar" alias is one of those. When usingan image URN, the validator fetches plan information and appends to the request, but this was missing when using an alias. Add the same handling to alias validation, so that aliases can be used as is without extra work from the user.

**Testing Guide**

```
g=group$RANDOM
az group create -g $g -l westeurope
az vm create -g $g -n vm$RANDOM --image Flatcar
```
Before this PR this results in an error ("missing plan information"), after this PR the result is success.

**History Notes**

[Compute] Fix #20174: `az vm create`: Determine plan information when using image alias

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [X] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [X] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [X] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
Compute

**Comment by wangzelin007** [comment]
Hi jepio,
Could you merge the latest code from remote dev branch first. After that, we will help you solve the remaining CI issue later.

**Comment by jepio** [comment]
@wangzelin007 OK, now I understand. I updated the dev branch now.

**Comment by wangzelin007** [comment]
Hi @jepio,
The real reason is that add _get_image_plan_info_if_exists will add one more api call.
All involved tests need to be re-recorded and re-upload
image

**Comment by jepio** [comment]
OK, how do I re-record the tests?

**Comment by wangzelin007** [comment]
> OK, how do I re-record the tests?
I am trying to re-record these tests, but many of then still failed in re-run.
image

get all failed test case in profile 2018-03-01-hybrid 
`azdev test vm --repo=./ --src=HEAD --tgt=origin/dev --no-exitfirst --profile 2018-03-01-hybrid --verbose --series`
re-run error cases to generate recording files:
`azdev test vm --no-exitfirst --live --lf --profile 2018-03-01-hybrid`

get all failed test case in profile 2019-03-01-hybrid
`azdev test vm --repo=./ --src=HEAD --tgt=origin/dev --no-exitfirst --profile 2019-03-01-hybrid --verbose --series`
re-run error cases to generate recording files:
`azdev test vm --no-exitfirst --live --lf --profile 2019-03-01-hybrid`

get all failed test case in profile 2020-09-01-hybrid
`azdev test vm --repo=./ --src=HEAD --tgt=origin/dev --no-exitfirst --profile 2020-09-01-hybrid --verbose --series`
re-run error cases to generate recording files:
`azdev test vm --no-exitfirst --live --lf --profile 2020-09-01-hybrid`

get all failed test case in profile latest
`azdev test vm --repo=./ --src=HEAD --tgt=origin/dev --no-exitfirst --profile latest --verbose --series`
re-run error cases to generate recording files:
`azdev test vm --no-exitfirst --live --lf --profile latest`

**Comment by wangzelin007** [comment]
ERROR INFO:
```json
{
  "status": "Failed",
  "error": {
    "code": "DeploymentFailed",
    "message": "At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/DeployOperations for usage details.",
    "details": [
      {
        "code": "BadRequest",
        "message": {
          "error": {
            "details": [
              {
                "message": "Could not find member vmSizeProperties on object of type VMHardwareProfile. Path properties.hardwareProfile.vmSizeProperties, line 1, position 110.",
                "target": "vm.properties.hardwareProfile.vmSizeProperties"
              },
              {
                "message": "Could not find member additionalCapabilities on object of type Properties. Path properties.additionalCapabilities, line 1, position 1230.",
                "target": "vm.properties.additionalCapabilities"
              }
            ],
            "code": "BadRequest",
            "message": "The request message is invalid."
          }
        }
      }
    ]
  }
}
```

File: src/azure-cli/azure/cli/command_modules/vm/_template_builder.py [hunk 1]
```diff
@@ -529,12 +532,13 @@ def _build_storage_profile():
...
-        vm_properties['additionalCapabilities']['ultraSSDEnabled'] = ultra_ssd_enabled
+    if cmd.supported_api_version(min_api='2018-06-01'):
+        vm_properties['additionalCapabilities'] = {}
+        if ultra_ssd_enabled is not None:
+            vm_properties['additionalCapabilities']['ultraSSDEnabled'] = ultra_ssd_enabled
 
-    if enable_hibernation is not None:
-        vm_properties['additionalCapabilities']['hibernationEnabled'] = enable_hibernation
+        if enable_hibernation is not None:
+            vm_properties['additionalCapabilities']['hibernationEnabled'] = enable_hibernation
```
Comment (wangzelin007) [inline src/azure-cli/azure/cli/command_modules/vm/_template_builder.py:541]: fix
```json
{
  "status": "Failed",
  "error": {
    "code": "DeploymentFailed",
    "message": "At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/DeployOperations for usage details.",
    "details": [
      {
        "code": "BadRequest",
        "message": {
          "error": {
            "details": [
              {
                "message": "Could not find member vmSizeProperties on object of type VMHardwareProfile. Path properties.hardwareProfile.vmSizeProperties, line 1, position 110.",
                "target": "vm.properties.hardwareProfile.vmSizeProperties"
              },
              {
                "message": "Could not find member additionalCapabilities on object of type Properties. Path properties.additionalCapabilities, line 1, position 1230.",
                "target": "vm.properties.additionalCapabilities"
              }
            ],
            "code": "BadRequest",
            "message": "The request message is invalid."
          }
        }
      }
    ]
  }
}
```

File: src/azure-cli/azure/cli/command_modules/vm/tests/hybrid_2018_03_01/test_vm_commands.py [hunk 2]
```diff
@@ -768,7 +768,7 @@ def test_vm_extension(self, resource_group):
         self.cmd('vm extension set -n {ext} --publisher {pub} --version 1.2  --vm-name {vm} --resource-group {rg} --protected-settings "{config}" --force-update')
         self.cmd('vm get-instance-view -n {vm} -g {rg}', checks=[
             self.check('*.extensions[0].name', ['VMAccessForLinux']),
-            self.check('*.extensions[0].typeHandlerVersion', ['1.4.7.1'])
+            self.check('*.extensions[0].typeHandlerVersion', ['1.5.11'])
```
Comment (wangzelin007) [inline src/azure-cli/azure/cli/command_modules/vm/tests/hybrid_2018_03_01/test_vm_commands.py:771]: fix

File: src/azure-cli/azure/cli/command_modules/vm/tests/hybrid_2018_03_01/test_vm_commands.py [hunk 3]
```diff
@@ -928,6 +928,7 @@ def test_vm_create_none_options(self, resource_group):
 
 class VMBootDiagnostics(ScenarioTest):
 
+    @AllowLargeResponse()
```
Comment (wangzelin007) [inline src/azure-cli/azure/cli/command_modules/vm/tests/hybrid_2018_03_01/test_vm_commands.py:931]: fix
AllowLargeResponse

File: src/azure-cli/azure/cli/command_modules/vm/_template_builder.py [hunk 4]
```diff
@@ -498,15 +498,18 @@ def _build_storage_profile():
...
     if v_cpus_available is not None:
         vm_size_properties['vCPUsAvailable'] = v_cpus_available
 
     if v_cpus_per_core is not None:
         vm_size_properties['vCPUsPerCore'] = v_cpus_per_core
 
-    vm_properties = {'hardwareProfile': {'vmSize': size, 'vmSizeProperties': vm_size_properties},
-                     'networkProfile': {'networkInterfaces': nics}, 'storageProfile': _build_storage_profile()}
+    if vm_size_properties:
+        vm_properties['hardwareProfile']['vmSizeProperties'] = vm_size_properties
```
Comment (wangzelin007) [inline src/azure-cli/azure/cli/command_modules/vm/_template_builder.py:512]: fix
```json
{
  "status": "Failed",
  "error": {
    "code": "DeploymentFailed",
    "message": "At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/DeployOperations for usage details.",
    "details": [
      {
        "code": "BadRequest",
        "message": {
          "error": {
            "details": [
              {
                "message": "Could not find member vmSizeProperties on object of type VMHardwareProfile. Path properties.hardwareProfile.vmSizeProperties, line 1, position 110.",
                "target": "vm.properties.hardwareProfile.vmSizeProperties"
              },
              {
                "message": "Could not find member additionalCapabilities on object of type Properties. Path properties.additionalCapabilities, line 1, position 1230.",
                "target": "vm.properties.additionalCapabilities"
              }
            ],
            "code": "BadRequest",
            "message": "The request message is invalid."
          }
        }
      }
    ]
  }
}
```

File: src/azure-cli/azure/cli/command_modules/vm/tests/hybrid_2018_03_01/test_vm_commands.py [hunk 5]
```diff
@@ -1148,7 +1149,7 @@ def test_vm_create_custom_ip(self, resource_group):
             'public_ip_sku': 'Standard'
         })
 
-        self.cmd('vm create -n {vm} -g {rg} --image openSUSE-Leap --admin-username user11 --private-ip-address 10.0.0.5 --public-ip-sku {public_ip_sku} --public-ip-address-dns-name {dns} --generate-ssh-keys')
+        self.cmd('vm create -n {vm} -g {rg} --image UbuntuLTS --admin-username user11 --private-ip-address 10.0.0.5 --public-ip-sku {public_ip_sku} --public-ip-address-dns-name {dns} --generate-ssh-keys')
```
Comment (wangzelin007) [inline src/azure-cli/azure/cli/command_modules/vm/tests/hybrid_2018_03_01/test_vm_commands.py:1152]: fix
ImageNotFound2

File: src/azure-cli/azure/cli/command_modules/vm/tests/hybrid_2018_03_01/test_vm_commands.py [hunk 6]
```diff
@@ -1159,7 +1160,7 @@ def test_vm_create_custom_ip(self, resource_group):
                  checks=self.check('ipConfigurations[0].privateIpAllocationMethod', 'Static'))
 
         # verify the default should be "Basic" sku with "Dynamic" allocation method
-        self.cmd('vm create -n {vm2} -g {rg} --image openSUSE-Leap --admin-username user11 --generate-ssh-keys')
+        self.cmd('vm create -n {vm2} -g {rg} --image UbuntuLTS --admin-username user11 --generate-ssh-keys')
```
Comment (wangzelin007) [inline src/azure-cli/azure/cli/command_modules/vm/tests/hybrid_2018_03_01/test_vm_commands.py:1163]: fix
ImageNotFound2

File: src/azure-cli/azure/cli/command_modules/vm/tests/hybrid_2018_03_01/test_vm_commands.py [hunk 7]
```diff
@@ -1283,6 +1284,7 @@ def test_vm_create_custom_data(self, resource_group):
 
 class VMSSCreateAndModify(ScenarioTest):
 
+    @AllowLargeResponse()
```
Comment (wangzelin007) [inline src/azure-cli/azure/cli/command_modules/vm/tests/hybrid_2018_03_01/test_vm_commands.py:1287]: same above

File: src/azure-cli/azure/cli/command_modules/vm/tests/hybrid_2018_03_01/test_vm_commands.py [hunk 8]
```diff
@@ -1497,7 +1499,7 @@ def test_vmss_accelerated_networking(self, resource_group):
         self.kwargs.update({
             'vmss': 'vmss1'
         })
-        self.cmd("vmss create -n {vmss} -g {rg} --vm-sku Standard_DS4_v2 --image Win2022Datacenter --admin-username clittester --admin-password Test12345678!!! --accelerated-networking --instance-count 1")
+        self.cmd("vmss create -n {vmss} -g {rg} --vm-sku Standard_DS4_v2 --image Win2012R2Datacenter --admin-username clittester --admin-password Test12345678!!! --accelerated-networking --instance-count 1")
```
Comment (wangzelin007) [inline src/azure-cli/azure/cli/command_modules/vm/tests/hybrid_2018_03_01/test_vm_commands.py:1502]: fix
Win2022Datacenter

File: src/azure-cli/azure/cli/command_modules/vm/tests/hybrid_2019_03_01/test_vm_commands.py [hunk 9]
```diff
@@ -1077,6 +1077,7 @@ def test_vm_create_none_options(self, resource_group):
 
 class VMBootDiagnostics(ScenarioTest):
 
+    @AllowLargeResponse()
```
Comment (wangzelin007) [inline src/azure-cli/azure/cli/command_modules/vm/tests/hybrid_2019_03_01/test_vm_commands.py]: same above

File: src/azure-cli/azure/cli/command_modules/vm/tests/hybrid_2019_03_01/test_vm_commands.py [hunk 10]
```diff
@@ -1343,7 +1344,7 @@ def test_vm_create_custom_ip(self, resource_group):
             'public_ip_sku': 'Standard'
         })
 
-        self.cmd('vm create -n {vm} -g {rg} --image openSUSE-Leap --admin-username user11 --private-ip-address 10.0.0.5 --public-ip-sku {public_ip_sku} --public-ip-address-dns-name {dns} --generate-ssh-keys')
+        self.cmd('vm create -n {vm} -g {rg} --image UbuntuLTS --admin-username user11 --private-ip-address 10.0.0.5 --public-ip-sku {public_ip_sku} --public-ip-address-dns-name {dns} --generate-ssh-keys')
```
Comment (wangzelin007) [inline src/azure-cli/azure/cli/command_modules/vm/tests/hybrid_2019_03_01/test_vm_commands.py]: same above

File: src/azure-cli/azure/cli/command_modules/vm/tests/hybrid_2019_03_01/test_vm_commands.py [hunk 11]
```diff
@@ -1354,7 +1355,7 @@ def test_vm_create_custom_ip(self, resource_group):
                  checks=self.check('ipConfigurations[0].privateIpAllocationMethod', 'Static'))
 
         # verify the default should be "Basic" sku with "Dynamic" allocation method
-        self.cmd('vm create -n {vm2} -g {rg} --image openSUSE-Leap --admin-username user11 --generate-ssh-keys')
+        self.cmd('vm create -n {vm2} -g {rg} --image UbuntuLTS --admin-username user11 --generate-ssh-keys')
```
Comment (wangzelin007) [inline src/azure-cli/azure/cli/command_modules/vm/tests/hybrid_2019_03_01/test_vm_commands.py]: same above

File: src/azure-cli/azure/cli/command_modules/vm/tests/hybrid_2019_03_01/test_vm_commands.py [hunk 12]
```diff
@@ -1490,6 +1491,7 @@ def test_vm_create_custom_data(self, resource_group):
 
 class VMSSCreateAndModify(ScenarioTest):
 
+    @AllowLargeResponse()
```
Comment (wangzelin007) [inline src/azure-cli/azure/cli/command_modules/vm/tests/hybrid_2019_03_01/test_vm_commands.py]: same above

File: src/azure-cli/azure/cli/command_modules/vm/tests/hybrid_2019_03_01/test_vm_commands.py [hunk 13]
```diff
@@ -2130,7 +2132,7 @@ def test_vm_error_on_zone_unavailable(self, resource_group, resource_group_locat
         try:
             self.cmd('vm create -g {rg} -n vm1 --admin-username clitester --admin-password PasswordPassword1! --image debian --zone 1')
         except Exception as ex:
-            self.assertTrue('availability zone is not yet supported' in str(ex))
+            self.assertTrue('does not support availability zones at location' in str(ex))
```
Comment (wangzelin007) [inline src/azure-cli/azure/cli/command_modules/vm/tests/hybrid_2019_03_01/test_vm_commands.py]: fix
AssertionError

File: src/azure-cli/azure/cli/command_modules/vm/tests/hybrid_2020_09_01/test_vm_commands.py [hunk 14]
```diff
@@ -905,6 +905,7 @@ def test_vm_create_existing_ids_options(self, resource_group, storage_account):
 
 class VMSSCreateAndModify(ScenarioTest):
 
+    @AllowLargeResponse()
```
Comment (wangzelin007) [inline src/azure-cli/azure/cli/command_modules/vm/tests/hybrid_2020_09_01/test_vm_commands.py]: same above

File: src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py [hunk 15]
```diff
@@ -3992,7 +3992,7 @@ def test_vm_error_on_zone_unavailable(self, resource_group, resource_group_locat
         try:
             self.cmd('vm create -g {rg} -n vm1 --admin-username clitester --admin-password PasswordPassword1! --image debian --zone 1')
         except Exception as ex:
-            self.assertTrue('availability zone is not yet supported' in str(ex))
+            self.assertTrue('does not support availability zones at location' in str(ex))
```
Comment (wangzelin007) [inline src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py]: same above

File: src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py [hunk 16]
```diff
@@ -6728,7 +6730,7 @@ def test_vmss_windows_patch_mode(self, resource_group):
             'rg': resource_group
         })
 
-        self.cmd('vmss create -g {rg} -n {vmss} --image Win2022Datacenter --enable-agent --enable-auto-update false --patch-mode Manual --orchestration-mode Flexible --admin-username azureuser --admin-password testPassword0')
+        self.cmd('vmss create -g {rg} -n {vmss} --image Win2012Datacenter --enable-agent --enable-auto-update false --patch-mode Manual --orchestration-mode Flexible --admin-username azureuser --admin-password testPassword0')
```
Comment (wangzelin007) [inline src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py]: same as above

File: src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py [hunk 17]
```diff
@@ -5367,31 +5367,33 @@ def test_dedicated_host_e2e(self, resource_group, resource_group_location):
         self.cmd('vm delete --name vm2 -g {rg2} --yes')
         self.cmd('vm host delete --name {host-name} --host-group {host-group} -g {rg2} --yes')
 
-    @ResourceGroupPreparer(name_prefix='cli_test_dedicated_host_', location='westus2')
+    @ResourceGroupPreparer(name_prefix='cli_test_dedicated_host3_', location='eastus2')
```
Comment (wangzelin007) [inline src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py]: fix
test_update_dedicated_host_e2e

File: src/azure-cli/azure/cli/command_modules/vm/tests/hybrid_2018_03_01/test_vm_commands.py (hunk 8 above)
Comment (zhoxing-ms) [reply src/azure-cli/azure/cli/command_modules/vm/tests/hybrid_2018_03_01/test_vm_commands.py:1502]: Using image a is by design, PR link: https://github.com/Azure/azure-cli/pull/20701, I think we need to investigate the reason why this image can no longer be found

**Comment by zhoxing-ms** [comment]
Replaced by #21666

**Pull request closed without merging**